FuncGodot Manual/pages/tips_runtime_building.html

```diff
@@ -0,0 +1,126 @@
+# Consider creating a procedural Voxel GI solid class entity that can be baked at runtime instead.
+func _build_complete() -> void:
+    print("Success! Unwrapping UV2...")
+    func_godot_map.unwrap_uv2()
+
+func _build_failed() -> void:
+    printerr("Failed to build the map file! :(")
+
+func _unwrap_uv2_complete() -> void:
+    print("UV2 unwrapping completed!")</code></pre>
+    <p>An example of this can be found in <a href="https://github.com/func-godot/func_godot_runtime_building" target="_blank">func_godot_runtime_building repository</a>.</p>
+
+    <h1>Building from Map Editors</h1>
+    <p>
+        Much like idTech engine games, it's possible to build and launch directly into a map from some map editors.
+        This can be quite a helpful tool for testing rapid iterations.
+    </p>
+
+    <h2>Godot Project Setup Example</h2>
+    <p>To build and launch our maps from the editor, you'll need to employ the <a href="https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-get-cmdline-user-args" target="_blank">OS class' ability to retrieve command line arguments</a>. A basic example might look something like this:</p>
+
+
+    <code><pre>
+# This script should be autoloaded to handle command line arguments
+# Additionally, requires a valid reference to a FuncGodot map
+
+var func_godot_map: FuncGodotMap
+
+var map_override: String
+var cmd_line_args := OS.get_cmdline_user_args()
+
+# Loop over all command line arguments, split on a delimiter for each command, and use this for the file path
+for cmd_line_arg in cmd_line_args:
+    if cmd_line_arg.contains("launch_debug::"):
+        var arg_and_value := cmd_line_arg.split("::")
+        # This should split into an array of size 2, containing the parameter name and its corresponding value
+        if arg_and_value.size() != 2:
+            return
+
+        map_override = arg_and_value[1]
+
+        func_godot_map.local_map_file = map_override
+        func_godot_map.verify_and_build()</code></pre>
+    <h2>Setting Up Your Editor</h2>
+    <p>In the below guides, there will be references made to executables. This is up to the user, however you can use the Godot executable to run your game straight from the editor, or you could use exported binaries of your game. On Windows, your Godot Binary path is where you keep your Godot.exe file. On most Linux systems this is found under <code>/usr/bin/godot</code>. An example executable path and parameters for running a game in the editor: <code class="highlight">C:/GAMEDEV/Godot.exe --path C:/func_godot_project_example/ -- launch_debug::${MAP_BASE_NAME}.map</code>. Note that <code class="highlight">--path</code> is to the directory where project.godot resides. See the full list of Godot engine <a href="https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html" target="_blank">Command Line Interface options</a>.</p>
```

Comment:
Personally I have not. From @RisingThumb I saw a TB path implying it could. I'll have to test myself, but used an exported project as an example for a reason. That may need to be specified if so.

I think one can do something like this and tack on other things, including running scripts as well-
per [this discussion](https://github.com/godotengine/godot-proposals/discussions/8664#discussioncomment-7871876)
`godot.exe --editor --headless res://SomeScene.tscn` is valid and if it's a tool script it can do things.

Reply:
Also no lightmap baking or UV2 unwrapping. :(

But yeah, I don't see a point for `Map Editor > Run Map through Godot Project` so I'm leaning towards "let's change the text in the tutorial to reflect that" for that section, to put an emphasis on "this is a process for exported projects".

Reply:
> doesn't match the example images. My first thought was "update the example images to reflect the text" but now I'm wondering if it's possible to run a scene through command line arguments that can run the launch options? Autoloads don't work at Editor Time as they don't get created until the start of Runtime.

This doesn't match because the example images are for an exported project. This is giving an example command you can use in a CLI for running from the editor.

> Have we run tests on whether you can do this from a non-exported project yet? Trying to figure out if we need to re-write the text or take new screenshots.

I have tested this out and it's what I currently do in my game. If you look at the commit before I used sinewave's screenshots, you'll see it reflected this and was what the screenshots showed. An example script for my game currently is this:

`/usr/bin/godot --path /home/odst/Programming/godot4/PerspectiveGame/ -- launch_debug::${MAP_BASE_NAME}.map`

> Autoloads don't work at Editor Time as they don't get created until the start of Runtime.

Depends what you mean by Editor Time. If you mean in the editor, yes this is correct. If you mean running from the editor, this is incorrect.

> Also no lightmap baking or UV2 unwrapping. :(

I've got a PR for Godot that should address this for running it from the editor. Exports don't work currently.

> For more context, at least in my use case, and what I think this is likely best applied to, it makes sense to primarly use this with exported projects.

Personally, I think it makes sense much more running from editor. You can bake lightmaps from the editor build(after the PR is merged upstream), and you can commit to version control more easily as you don't keep maps outside of the export, and also with the other work that Psprite was doing with extending the imports so you can just have the scene be an import, actually makes it very usable this way. This means that runtime map building can be used as an editor-only development aid, as opposed to an actively supported feature.

I think running from exported projects is much more specific to modders as you lose a lot of the flexibility of the editor, and the maps have to be bundled separately or copied over.

Regardless, there's a case to be made for both. I think we should signpost that both are options that can be taken, but it's up to people which works with their workflow best.

> would there even be a point to have a tutorial for non-exported projects working with map editors?

You can't do everything you might want to do in an exported project. You can't bake lightmaps, you can't change res:// files accordingly(which you might want to do for pregenerating stuff) etc. There's enough of a case for both that we shouldn't be opinionated about it.

Additionally, I don't think we should be making a tutorial for both non-exported and exported project. Saying one exists and an example should be sufficient. This is just demonstrating that you can run any executable from a map editor and doing a tutorial for this.

Reply:
> This doesn't match because the example images are for an exported project.

I know that. That's why I brought it up. Less concerned about a chicken and egg question and more concerned about "which direction should we take it?"

> Depends what you mean by Editor Time. If you mean in the editor, yes this is correct. If you mean running from the editor, this is incorrect.

No, when I say Editor Time and Runtime, I mean what is defined as Editor Time and Runtime by Godot. If you're running your code in the Editor, that's Editor Time. If you're running your code in the application, that's Runtime, regardless if it's F5 in Godot or running the exe.

> I've got a PR for Godot that should address this for running it from the editor. Exports don't work currently.

I was talking specifically about exports in response to @sinewavey . UV2 unwrapping through code does work in Editor. I'm aware that lightmap baking does not. Voxel GI technically can be baked from script, but it's buggy.

> Personally, I think it makes sense much more running from editor...

> Regardless, there's a case to be made for both. I think we should signpost that both are options that can be taken, but it's up to people which works with their workflow best.

> Additionally, I don't think we should be making a tutorial for both non-exported and exported project...

These are all points that make me lean towards writing the tutorial and making the image examples as the non-export project method, since the main difference between project and exe are that you point to the exported exe instead of godot.exe and you can omit the project path from the parameters. The fact that there's a little more involved in setting up the project method is why I _would_ lean towards that being the examples and the tutorial. The exported exe information and its limitations would fit more as a pair of noteblocks in this case, since most everything else about the process would be the same. 

> You can bake lightmaps from the editor build(after the PR is merged upstream)

This doesn't matter until it's actually merged in a Stable build or will be by the time this tutorial gets merged with the docs proper.

> you can commit to version control more easily as you don't keep maps outside of the export

This is fixed with a recent update I pushed. While the original .map files disappear, you still use `"res://mymap.map"` to access the map resource and the actual map data is retained now. There is no practical difference between the original map file and the QuakeMapFile resource: all of the data is retained. The only thing you'd have to do different is if you're doing a file check for a map file using `FileAccess.file_exists()`, at which point you'll have to make sure to check for both `.map` and `.map.import`. But that's just general Godot pattern stuff, if you just pass the expected map path through to FuncGodot it already performs both of those checks.

You can see the fix here:
https://github.com/func-godot/func_godot_plugin/commit/e711865f0b9343beb37f7bdd7b7226f8914027e2

Reply:
> If you're running your code in the Editor, that's Editor Time. If you're running your code in the application, that's Runtime, regardless if it's F5 in Godot or running the exe.

Ok, I was just confused. Yeah, idk if it's possible for the editor. You should be able to open the editor from it, but how CLI parameters are read, idk. Maybe an @tool script could handle it.

What I'm currently doing is opening it up as in running from the editor, so I can still modify res:// files

> These are all points that make me lean towards writing the tutorial and making the image examples as the non-export project method, since the main difference between project and exe are that you point to the exported exe instead of godot.exe and you can omit the project path from the parameters.

I agree. I can go ahead and revert the changes for the trenchbroom screenshots. @sinewavey would you also be able to provide a screenshot for this in NetRadiant?

> This is fixed with a recent update I pushed.

I was referring to where you put .map files when you only have an exported project. I think @sinewavey is using an external "SDK" like structure where the textures and maps are bundled outside of the exported project? Or at least that was something I saw them doing at one point.

Reply:
> Yeah, idk if it's possible for the editor. You should be able to open the editor from it, but how CLI parameters are read, idk. Maybe an @tool script could handle it.

Nah, even as a `@tool` script, an autoload won't run just because it doesn't exist. You could probably set up a scene with one of the nodes having a `@tool` script, but admittedly at that point we're also no longer doing Runtime building and need a different tutorial.

> I think @sinewavey is using an external "SDK" like structure where the textures and maps are bundled outside of the exported project?

Yeah, she was doing that partly because there were issues when trying to build from a packed map resource. I dunno if she's doing things any differently now though? In any case, it works both ways just fine now.

> If we're all in agreement, I can go ahead and revert the changes for the trenchbroom screenshots(nvm, new ones are needed for the command name above)

You can use this for the TrenchBroom image.
![image](https://github.com/func-godot/func_godot_docs/assets/44485952/038b6a04-7e82-4694-ad0b-2b1b4a874602)

Reply:
Hi, been a little busy today. I can provide fresh screenshots no problem. 

I was using an sdk like structure as my ultimate goal was to offer this for modders. Will follow up a bit more later, for now my distro to mappers is still the same overall process

Reply:
> Nah, even as a @tool script, an autoload won't run just because it doesn't exist. You could probably set up a scene with one of the nodes having a @tool script, but admittedly at that point we're also no longer doing Runtime building and need a different tutorial.

I meant that you could probably do the cli arguments parsing in a tool script instead of an autoloaded script for use with the editor, but idk how useful that is and is outside the scope as you said

Reply:
![image](https://github.com/func-godot/func_godot_docs/assets/90056029/decbf8fb-25d3-4872-b529-1783f4e82200)

```
<?xml version="1.0"?>
<project version="2.0">
	<var name="game">"[EnginePath]example_project.[ExecutableType]"</var>
	<build name="Run and build map in exported project">
		<command>[game] -- run_map::"[MapFile]"</command>
	</build>
</project>
```



I can now also confirm it does work with direct path and a tool script (I was curious and wanted to set this up for myself regardless, it'll help my mappers anyway to deal with needing to open Godot for the moment - I can really baby it and set stuff up for them. But again, with such limited scope, this is likely more of a personal detail that I imagine most larger projects can figure out

Variety of options below anyway from my experimenting


```
## This script is attached to a node with a FuncGodotMap child
## It is saved in scenes/map_builder.tscn
@tool
extends Node

@export var func_godot_map: FuncGodotMap


func _ready() -> void:
	if !func_godot_map:
		return

	var map_override: String
	var cmd_line_args := OS.get_cmdline_user_args()

	# Loop over all command line arguments, split on a delimiter for each command, and use this for the file path
	for cmd_line_arg in cmd_line_args:
		if cmd_line_arg.contains("run_map::"):

			# This should split into an array of size 2, containing the parameter name and its corresponding value
			var arg_and_value := cmd_line_arg.split("::")

			if arg_and_value.size() != 2:
				return

			map_override = arg_and_value[1]
			func_godot_map.local_map_file = map_override
			func_godot_map.verify_and_build()
			return

	return
```
	
The accompanying build script

```<?xml version="1.0"?>
<project version="2.0">
	<var name="game">"[EnginePath]example_project.[ExecutableType]"</var>
	<var name="godot">"C:\godot_path\Godot_v4.2.2-stable_win64.exe" -e --PATH "C:\func_godot\project.godot"</var>
	<build name="Run and build map in exported project">
		<command>[game] -- run_map::"[MapFile]"</command>
	</build>
	<separator/>
	<build name="Run and build map in Godot Editor">	
		<command>[godot] scenes\map_builder.tscn -- run_map::"[MapFile]"</command>
	</build>
</project>
```




![image](https://github.com/func-godot/func_godot_docs/assets/90056029/05f6176e-0be6-4905-9fe3-76dd9c680bac)
![image](https://github.com/func-godot/func_godot_docs/assets/90056029/20a2622e-8aa0-4440-8208-2facb614d05f)